openlibrary/plugins/openlibrary/js/ile/utils/SelectionManager/SelectionManager.js

```diff
@@ -82,19 +84,77 @@ export default class SelectionManager {
             ($(clickEvent.target).closest('a').is('a') &&
             $(clickEvent.target).not('.ile-select-handle').length > 0)) return;
 
         const el = clickEvent.currentTarget;
+        if (clickEvent.shiftKey && this.lastClicked)
+        {
+            this.clearTextSelection();
+            const siblingSet = this.getSiblingSet(el);
+            const lastClickedIndex = siblingSet.index(this.lastClicked);
+            const elIndex = siblingSet.index(el);
+            if (lastClickedIndex > -1 && Math.abs(elIndex - lastClickedIndex) > 1) {
+                let affectedElements;
+                if (elIndex > lastClickedIndex) {
+                    affectedElements = siblingSet.slice(lastClickedIndex + 1, elIndex + 1);
+                } else {
+                    affectedElements = siblingSet.slice(elIndex, lastClickedIndex);
+                }
+                const stateChange = el.classList.contains('ile-selected') ? false : true;
+                for (const element of affectedElements) this.toggleSelected(element, stateChange);
+            } else {
+                this.toggleSelected(el);
+            }
+        }
+        else {
+            this.toggleSelected(el);
+        }
+        this.lastClicked = el;
+        this.updateToolbar();
+    }
+
+    /**
+     * Sets of selectable elements are sometimes HTML siblings and sometimes not. This function
+     * hides that complexity by finding the common parent between the passed element and the last
+     * clicked element and generating a set of siblings from that information
+     *
+     * @param {HTMLElement} clicked
+     */
+    getSiblingSet(clicked) {
+        const lastParents = $(this.lastClicked).parents();
+        const clickedParents = $(clicked).parents();
+        let commonParent = undefined;
+        for (let i = 0; i < 2 && i < lastParents.length; i++) {
+            if (clickedParents.is(lastParents[i])) {
+                commonParent = lastParents[i];
+                break;
+            }
+        }
+        if (commonParent) {
+            return $(commonParent).find('.ile-selectable');
+        } else {
+            return $(clicked);
+        }
+    }
+
+    /**
+     * @param {HTMLElement} el
+     * @param {boolean} [forceSelected]
+     * If included, turns the toggle into a one way-only operation. If set to false, elements will only
+     * be deselected, not selected. If set to true, elements will only be selected, but not deselected.
+     */
+    toggleSelected(el, forceSelected) {
         const isCurSelected = el.classList.contains('ile-selected');
         const provider = this.getProvider(el);
         const olid = provider.data(el);
         const img_src = this.getType(olid)?.image(olid);
-        this.setElementSelectionAttributes(el, !isCurSelected);
 
+        if ((forceSelected === true && isCurSelected) || (forceSelected === false && !isCurSelected)) return;
+        this.setElementSelectionAttributes(el, !isCurSelected);
```

Comment:
That boolean logic is a touch confusing; 
```suggestion
        this.setElementSelectionAttributes(el, forceSelected == null ? !isCurSelected : forceSelected);
```

And set the default to be `null`

Reply:
Your solution is doing something different (that doesn't work :)  My confusing logic is there to handle any elements in the range that don't need to be modified because they're already flipped the way they should be. In that case, we just want to bail out of the method because otherwise the element will be incorrectly double-added or removed from the selection set.

The logic is IF (element state is already correct) return ELSE do all the toggle operations.

After staring at it for a bit, I realized I could simplify the test to `if (isCurSelected === forceSelected) return;` which may be a little less obscure.

Reply:
Nice! Ideally we'd make `setElementSelectionAttributes` work when called multiple times, but considering that out of scope of this pr!